### ISSUE TITLE
Gracefully shutdown subscriber

### ISSUE THREAD

**Issue by eladbitton** [googleapis/python-pubsub#276]
Hey,

#### Environment details

  - OS type and version: Ubuntu 18.04
  - Python version: 3.6.9
  - pip version: 9.0.1
  - `google-cloud-pubsub` version: 2.2.0

#### Code example

Sample code for a subscriber from Google's guide with minor changes that will show my issue:
```python

from concurrent.futures import TimeoutError
from google.cloud import pubsub_v1
from time import sleep

# TODO(developer)
# project_id = "your-project-id"
# subscription_id = "your-subscription-id"
# Number of seconds the subscriber should listen for messages
# timeout = 5.0

subscriber = pubsub_v1.SubscriberClient()
# The `subscription_path` method creates a fully qualified identifier
# in the form `projects/{project_id}/subscriptions/{subscription_id}`
subscription_path = subscriber.subscription_path(project_id, subscription_id)

def callback(message):
    print(f"Received {message}.")
    sleep(30)
    message.ack()
    print("Acked")

streaming_pull_future = subscriber.subscribe(subscription_path, callback=callback)
print(f"Listening for messages on {subscription_path}..\n")

# Wrap subscriber in a 'with' block to automatically call close() when done.
with subscriber:
    sleep(10)
    streaming_pull_future.cancel()
    streaming_pull_future.result()
```
From https://cloud.google.com/pubsub/docs/pull

I expect this code to stop pulling messages and finish the running messages and then exits.

Actually this code stops pulling messages and finish executing the running messages but it does not ack the messages. The .ack() happens but the server does not receive the ack, so next run the same messages return again.

1. Why doesn't the server receives the ack?

2. How to gracefully shutdown the subscriber?

3. What is the expected behavior of .cancel()?

Corresponding stackoverflow question:
https://stackoverflow.com/questions/65761254/pub-sub-python-client-gracefully-shutdown-subscriber

Thank you!

**Comment by plamut** [comment]
@eladbitton Thank you for the question. Please see the answers below and let us know if any additional clarification is needed, thanks!

> Why doesn't the server receives the ack?

It appears that the message callback sleeps for 30 seconds, but the shutdown is triggered after 10 seconds, and the processing stops.

The streaming pull is managed in the background by a `StreamingPullManager` class and on shutdown its [close()](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L516-L566) method is called. The method shuts down all helper threads, as well as the scheduler used for dispatching the messages to the user callback.

When scheduler [shutdown](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/scheduler.py#L113-L124) is triggered, all pending items from its work queue are discarded in a non-blocking fashion. The shutdown does not wait for the callbacks to complete, as this can potentially block for extended periods of time.

> How to gracefully shutdown the subscriber?

It's by cancelling the streaming pull future, which [invokes](https://github.com/googleapis/python-pubsub/blob/542d79d7c5fb7403016150ba477485756cd4097b/google/cloud/pubsub_v1/subscriber/futures.py#L46-L51) the abovementioned `close()` method on the streaming pull manager.

Since streaming pull is a long-running (possibly indefinite) operation, we don't want the main thread to exit prematurely. The canonical way of blocking is by calling `streaming_pull_future.result()` and shutting the future down on, say, a keyboard interrupt:
```py
try:
    streaming_pull_future.result()
except KeyboardInterrupt:
    streaming_pull_future.cancel()
```

Or to shut down after a pre-determined time period:

```py
try:
    streaming_pull_future.result(timeout=123)
except concurrent.futures.TimeoutError:
    streaming_pull_future.cancel()
```


> What is the expected behavior of .cancel()?

The cancel should stop the message stream in the background, gracefully shut down all background helper threads, and call any on close callbacks registered. If interested, you check the `close()` method for details.

**Comment by plamut** [comment]
Based on the first post and the comment on the linked StackOverflow question, I am re-classifying this as a feature request similar to e.g. #150.

**Comment by plamut** [comment]
Just to give a quick update - it turned out that when shutting down the scheduler, the streaming pull manager internally calls `shutdown()` [without any arguments](https://github.com/googleapis/python-pubsub/blob/8751bcc5eb782df55769b48253629a3bde3d4661/google/cloud/pubsub_v1/subscriber/scheduler.py#L124), and the shutdown actually **blocks** [by default](https://docs.python.org/3/library/concurrent.futures.html?highlight=futures#concurrent.futures.Executor.shutdown) until the currently executing callbacks are done executing.

However, the reason why `ack()` calls in callbacks had no effect is probably because the leaser thread does not wait for the scheduler shutdown to complete and instead exits long before the callbacks wake up and call `message.ack()`. The latter puts an AckRequest into a queue, but the message's ACK timeout (initially 10 seconds) already expires, as there is no running leaser anymore to automatically extend it.

TL; DR - the leaser thread should not exit when the manager becomes "inactive", i.e. when the background consumer thread is shut down, but should instead wait to be shut down explicitly by the manager, i.e. _after_ the scheduler is done shutting down.

---

I also noticed that the long sleep in callback actually blocks `streaming_pull_future.cancel()` call, though that might be undesirable for some users. We could make this behavior configurable along the way.

**Comment by plamut** [comment]
@eladbitton Apart from the threads exiting prematurely on shutdown and ACKs not being sent, it turned out that there's another possible reason why a message could be re-delivered.

For the ACKs to be persisted, the backend currently requires that **all** messages that were published in the same batch are ACK-ed. If some of the messages get ACK-ed but then the streaming pull is canceled, the entire batch of messages will be re-delivered.
